zoidberg/poloidal_grid.py

```diff
@@ -473,42 +469,48 @@ def metric(self):
 
         """
 
-        dx = 1.0 / float(self.nx - 1)  # x from 0 to 1
-        dz = 2.0 * np.pi / float(self.nz)  # z from 0 to 2pi
-
         # Get arrays of indices
         xind, zind = np.meshgrid(np.arange(self.nx), np.arange(self.nz), indexing="ij")
 
         # Calculate the gradient along each coordinate
-        dRdx, dZdx = self.getCoordinate(xind, zind, dx=1)
-        dRdx /= dx
-        dZdx /= dx
-        dRdz, dZdz = self.getCoordinate(xind, zind, dz=1)
-        dRdz /= dz
-        dZdz /= dz
-
-        g_xx = dRdx**2 + dZdx**2
-        g_xz = dRdx * dRdz + dZdx * dZdz
-        g_zz = dRdz**2 + dZdz**2
-
-        # Calculate metric by inverting
-        # ( gxx   gxz ) = ( g_xx   g_xz )^-1
-        # ( gxz   gzz )   ( g_xz   g_zz )
-
-        determinant = g_xx * g_zz - g_xz**2
-        gxx = g_zz / determinant
-        gzz = g_xx / determinant
-        gxz = -g_xz / determinant
+        dolddnew = np.array(
+            [self.getCoordinate(xind, zind, dx=a, dz=b) for a, b in ((1, 0), (0, 1))]
+        )
+        # dims: 0 : dx/dz  - 1 : R/z - 2,3 : spatial (r, \theta)
+        ddist = np.sqrt(np.sum(dolddnew**2, axis=1))  # sum R + z
+        sumdz = np.sum(ddist[1], axis=1)  # sum in r direction
+        ddist[1] *= 2 * np.pi / sumdz[..., None]  # normalise theta (0 -> 2 pi)
+        # Transform derivatives from index space to real space.
+        dolddnew /= ddist[:, None, ...]
+
+        # g_ij = J_ki J_kj
+        # (2.5.27) from D'Haeseleer 1991
+        # Note: our J is transposed
+        J = dolddnew
+        g = np.sum(
+            np.array(
+                [
+                    [[J[j, i] * J[k, i] for i in range(2)] for j in range(2)]
+                    for k in range(2)
+                ]
+            ),
+            axis=2,
+        )
 
+        assert np.all(g[0, 0] > 0)
+        assert np.all(g[1, 1] > 0)
```

Comment:
@ZedThree would that be helpful:
```suggestion
        assert np.all(g[0, 0] > 0), f"g[0, 0] is expected to be positive, but some values are not (minimum {np.min(g[0, 0])})"
        assert np.all(g[1, 1] > 0), f"g[1, 1] is expected to be positive, but some values are not (minimum {np.min(g[1, 1])})"
```

Reply:
Perfect, thanks!